### PR TITLE
glib: Improve the API of Type

### PR DESCRIPTION
**Pull request by heftig**
- Make `Type` leaner, replace enum with assoc consts.
- Pass `Type` by value, where possible
- Introduce `Type::is_valid` and use it
- Make `Type::name` return a `str`

### CODE REVIEW

**Comment by sdroege** [comment]
This kind of handles https://github.com/gtk-rs/gtk-rs/issues/313 (for `Type`) at least, which is good.

File: glib/src/types.rs [hunk 1]
```diff
@@ -13,70 +13,89 @@ use std::mem;
...
     /// The fundamental type from which all objects are derived
-    Object,
-    /// A non-fundamental type identified by value of type `usize`
-    Other(usize),
-}
+    pub const OBJECT: Self = Self(gobject_ffi::G_TYPE_OBJECT);
 
-impl Type {
     #[doc(alias = "g_type_name")]
     pub fn name(&self) -> String {
```
Comment (sdroege) [inline glib/src/types.rs:87]: Can you make this

```rust
     pub fn name<'a>(&'a self) -> &'a str { 
```
in a separate commit here? The name is backed by a quark so it's by definition a static reference

File: glib/src/types.rs [hunk 2]
```diff
@@ -100,7 +100,7 @@ impl Type {
     }
 
     #[doc(alias = "g_type_is_a")]
-    pub fn is_a(&self, other: &Type) -> bool {
+    pub fn is_a(&self, other: Self) -> bool {
```
Comment (sdroege) [inline glib/src/types.rs:103]: Should do the same for all the other functions here, I guess. Take `self` instead of `&self`, etc.

**Review by sdroege — COMMENTED** [review]
Looks good to me otherwise, thanks! That should improve the code generated by the compiler for these things quite a bit and keep it simpler.

File: glib/src/types.rs (hunk 2 above)
Comment (heftig) [reply glib/src/types.rs:103]: Leaving `is_valid` with `&self` made it possible to write `.filter(Type::is_valid)`. On the other hand, `map` would prefer `self`.

Comment (sdroege) [reply glib/src/types.rs:103]: The reasoning here is mostly https://rust-lang.github.io/rust-clippy/master/index.html#trivially_copy_pass_by_ref

Comment (heftig) [reply glib/src/types.rs]: Done.

File: glib/src/types.rs (hunk 1 above)
Comment (heftig) [reply glib/src/types.rs]: Done.

**Review by sdroege — APPROVED** [review]

**Review by bilelmoussaoui — APPROVED** [review]

**Pull request merged**